be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ParamsWithMapHandler.scala

```diff
@@ -0,0 +1,133 @@
+
+        val handlerMerger = builder.add(Merge[GraphMessage](totalPorts))
+
+        /* glue the components together */
+        handlerPartitioner.out(portPipelineError) ~> handlerMerger
+        handlerPartitioner.out(portHeartBeatHandler) ~> heartbeatHandler ~> handlerMerger
+        handlerPartitioner.out(portGetMsgsByIdHandler) ~> getMessagesByIdHandler ~> handlerMerger
+
+        /* close the shape */
+        FlowShape(handlerPartitioner.in, handlerMerger.out)
+
+      }
+  })
+
+  private def heartbeatHandler(dbActorRef: AskableActorRef): Flow[GraphMessage, GraphMessage, NotUsed] = Flow[GraphMessage].map({
+    case Right(jsonRpcMessage: JsonRpcRequest) =>
+      /** first step is to retrieve the received heartbeat from the jsonRpcRequest * */
+      val receivedHeartBeat: Map[Channel, Set[Hash]] = jsonRpcMessage.getParams.asInstanceOf[Heartbeat].channelsToMessageIds
+
+      /** second step is to retrieve the local set of channels * */
+      var setOfChannels: Set[Channel] = Set()
+      val ask = dbActorRef ? DbActor.GetAllChannels()
+      Await.ready(ask, duration).value match {
+        case Some(Success(DbActor.DbActorGetAllChannelsAck(channels))) =>
+          setOfChannels = channels
+
+        case Some(Failure(ex: DbActorNAckException)) => Left(PipelineError(ex.code, s"couldn't retrieve local set of channels", jsonRpcMessage.getId))
+        case reply                                   => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, s"heartbeatHandler failed : unexpected DbActor reply '$reply'", jsonRpcMessage.getId))
+      }
+
+      /** third step is to ask the DB for the content of each channel in terms of message ids. * */
+      val localHeartBeat: mutable.HashMap[Channel, List[Hash]] = mutable.HashMap()
+      setOfChannels.foreach(channel => {
+        val ask = dbActorRef ? DbActor.ReadChannelData(channel)
+        Await.ready(ask, duration).value match {
+          case Some(Success(DbActor.DbActorReadChannelDataAck(channelData))) =>
+            val setOfIds = channelData.messages.to(collection.immutable.Set)
+            localHeartBeat += (channel -> setOfIds.toList)
+          case Some(Failure(ex: DbActorNAckException)) => Left(PipelineError(ex.code, s"couldn't readChannelData for local heartbeat", jsonRpcMessage.getId))
+          case reply                                   => Left(PipelineError(ErrorCodes.SERVER_ERROR.id, s"heartbeatHandler failed : unexpected DbActor reply '$reply'", jsonRpcMessage.getId))
+        }
+      })
+
+      /** finally, we only keep from the received heartbeat the message ids that are not contained in the locally extracted heartbeat. * */
+      val missingIds: mutable.HashMap[Channel, Set[Hash]] = mutable.HashMap()
+      receivedHeartBeat.keys.foreach(channel => {
+        if (localHeartBeat.contains(channel)) {
+          missingIds += (channel -> receivedHeartBeat(channel).filter(id =>
+            !localHeartBeat(channel).contains(id)
```

Comment:
If you take the previous comment into account, there is a nice `.diff` function that you can use to get the difference between sets